dev/ese/src/sync/sync.cxx

```diff
@@ -4303,6 +3817,323 @@ void CKernelSemaphore::Release( const INT cToRelease )
 }
 
 
+//  Semaphore
+
+//  ctor
+
+CSemaphore::CSemaphore( const CSyncBasicInfo& sbi )
+    :   CEnhancedStateContainer< CSemaphoreState, CSyncStateInitNull, CSemaphoreInfo, CSyncBasicInfo >( syncstateNull, sbi )
+{
+    //  further init of CSyncBasicInfo
+
+    State().SetTypeName( "CSemaphore" );
+    State().SetInstance( (CSyncObject*)this );
+}
+
+//  dtor
+
+CSemaphore::~CSemaphore()
+{
+#ifdef SYNC_ANALYZE_PERFORMANCE
+#ifdef SYNC_DUMP_PERF_DATA
+
+    //  dump performance data
+
+    OSSyncStatsDump(    State().SzTypeName(),
+                        State().SzInstanceName(),
+                        State().Instance(),
+                        (DWORD)-1,
+                        State().CWaitTotal(),
+                        State().CsecWaitElapsed(),
+                        State().CAcquireTotal(),
+                        State().CContendTotal(),
+                        0,
+                        0 );
+
+#endif  //  SYNC_DUMP_PERF_DATA
+#endif  //  SYNC_ANALYZE_PERFORMANCE
+}
+
+//  converts the internal timeout value to an OS level timeout
+
+const DWORD CSemaphore::_DwOSTimeout( const INT cmsecTimeout )
+{
+    if ( cmsecTimeout == cmsecInfinite || cmsecTimeout == cmsecInfiniteNoDeadlock )
+    {
+        return INFINITE;
+    }
+    else if ( cmsecTimeout >= 0 )
+    {
+        return cmsecTimeout;
+    }
+    else
+    {
+        return 0;
+    }
+}
+
+//  attempts to acquire a count from the semaphore, returning fFalse if unsuccessful
+//  in the time permitted
+
+const BOOL CSemaphore::_FAcquire( const DWORD dwTimeout )
+{
+    const DWORD dwStart = DwOSSyncITickTime();
+    DWORD dwRemaining = dwTimeout;
+
+    //  try forever until we successfully change the state of the semaphore
+
+    State().IncWait();
+    OSSYNC_FOREVER
+    {
+        //  read the current state of the semaphore
+
+        const CSemaphoreState stateCur = (CSemaphoreState&) State();
+
+        //  see if we have an available count
+
+        if ( stateCur.CAvail() > 0 )
+        {
+            //  we ourselves are waiting on the semaphore, so there had better be at least
+            //  one waiter in the state
+
+            OSSYNCAssert( stateCur.CWait() > 0 );
+
+            //  try to atomically acquire the semaphore and decrement the number of waiters
```

Comment:
It seems that the current state may change between the times when it is checked, decremented, and finally exchanged. It looks to me that the value we read at line 3890 could potentially be changed in, e.g., CSemaphore::_FTryAcquire/CSemaphoreState::FDecAvail. At the very least, it's not obvious to me why stateCur should remain unchanged between these calls.

Reply:
Actually, I just tested it in a small example, and it seems to work (the read state is copied), so that means that the atomic comparison and the decrementing should be valid.

Reply:
Yes, the state is read once and is stored in the local variable.

So while the actual state of the semaphore may change, we would still be working with that local copy. And that copy would be passed as the compare value [during CAS in line 3905](https://github.com/microsoft/Extensible-Storage-Engine/blob/0b755685b5d6702c1a24b8a4b4b86f41d3ed7124/dev/ese/src/sync/sync.cxx#L3905). If the actual state did change, the CAS is going to fail, so we will read the new state and try changing it on the next iteration of the loop.

Looking at the [original version of sync.cxx](https://github.com/microsoft/Extensible-Storage-Engine/blob/3bcf428c8a041733043e18fd9ae55cffeba307b2/dev/ese/src/sync/sync.cxx#L435), I think that the code before the change worked similarly in this aspect.